File: frontend/src/mocks/handlers/report-history.ts
/**
 * Report History Mock Handlers
 *
 * MSW handlers for report history API endpoints.
 */
import { http, HttpResponse, delay } from 'msw'
import {
  createGeneratedReport,
  generateReports,
  generateReportStatistics,
  type GeneratedReport,
  type ReportStatistics,
} from '../factories/reports'

const API_BASE = '/api/v1'

// In-memory store for generated reports
let reportsStore: GeneratedReport[] = generateReports(30)

/**
 * Get filtered reports based on query parameters.
 */
function filterReports(params: URLSearchParams): { items: GeneratedReport[]; total: number } {
  let filtered = [...reportsStore]

  // Apply filters
  const sourceId = params.get('source_id')
  if (sourceId) {
    filtered = filtered.filter((r) => r.source_id === sourceId)
  }

  const validationId = params.get('validation_id')
  if (validationId) {
    filtered = filtered.filter((r) => r.validation_id === validationId)
  }

  const reporterId = params.get('reporter_id')
  if (reporterId) {
    filtered = filtered.filter((r) => r.reporter_id === reporterId)
  }

  const format = params.get('format')
  if (format) {
    filtered = filtered.filter((r) => r.format === format)
  }

  const status = params.get('status')
  if (status) {
    filtered = filtered.filter((r) => r.status === status)
  }

  const includeExpired = params.get('include_expired') === 'true'
  if (!includeExpired) {
    filtered = filtered.filter((r) => r.status !== 'expired')
  }

  const search = params.get('search')
  if (search) {
    const lowerSearch = search.toLowerCase()
    filtered = filtered.filter((r) => r.name.toLowerCase().includes(lowerSearch))
  }

  // Sorting
  const sortBy = params.get('sort_by') || 'created_at'
  const sortOrder = params.get('sort_order') || 'desc'
  filtered.sort((a, b) => {
    const aVal = (a as unknown as Record<string, unknown>)[sortBy]
    const bVal = (b as unknown as Record<string, unknown>)[sortBy]
    if (aVal === bVal) return 0
    if (sortOrder === 'desc') {
      return aVal && bVal && aVal > bVal ? -1 : 1
    }
    return aVal && bVal && aVal < bVal ? -1 : 1
  })

  // Pagination
  const page = parseInt(params.get('page') || '1', 10)
  const pageSize = parseInt(params.get('page_size') || '20', 10)
  const start = (page - 1) * pageSize
  const items = filtered.slice(start, start + pageSize)

  return {
    items,
    total: filtered.length,
  }
}

export const reportHistoryHandlers = [
  // List reports
  http.get(`${API_BASE}/reports/history`, async ({ request }) => {
    await delay(300)
    const url = new URL(request.url)
    const { items, total } = filterReports(url.searchParams)
    const page = parseInt(url.searchParams.get('page') || '1', 10)
    const pageSize = parseInt(url.searchParams.get('page_size') || '20', 10)

    return HttpResponse.json({
      items,
      total,
      page,
      page_size: pageSize,
    })
  }),

  // Get statistics
  http.get(`${API_BASE}/reports/history/statistics`, async () => {
    await delay(200)
    const stats = generateReportStatistics(reportsStore)
    return HttpResponse.json(stats)
  }),

  // Get single report
  http.get(`${API_BASE}/reports/history/:reportId`, async ({ params }) => {
    await delay(150)
    const { reportId } = params
    const report = reportsStore.find((r) => r.id === reportId)

    if (!report) {
      return HttpResponse.json({ detail: 'Report not found' }, { status: 404 })
    }

    return HttpResponse.json(report)
  }),

  // Create report record
  http.post(`${API_BASE}/reports/history`, async ({ request }) => {
    await delay(300)
    const data = (await request.json()) as Partial<GeneratedReport>
    const newReport = createGeneratedReport({
      ...data,
      status: 'pending',
    })
    reportsStore.unshift(newReport)
    return HttpResponse.json(newReport, { status: 201 })
  }),

  // Update report record
  http.patch(`${API_BASE}/reports/history/:reportId`, async ({ params, request }) => {
    await delay(200)
    const { reportId } = params
    const index = reportsStore.findIndex((r) => r.id === reportId)

    if (index === -1) {
      return HttpResponse.json({ detail: 'Report not found' }, { status: 404 })
    }

    const updates = (await request.json()) as Partial<GeneratedReport>
    reportsStore[index] = {
      ...reportsStore[index],
      ...updates,
      updated_at: new Date().toISOString(),
    }

    return HttpResponse.json(reportsStore[index])
  }),

  // Delete report record
  http.delete(`${API_BASE}/reports/history/:reportId`, async ({ params }) => {
    await delay(200)
    const { reportId } = params
    const index = reportsStore.findIndex((r) => r.id === reportId)

    if (index === -1) {
      return HttpResponse.json({ detail: 'Report not found' }, { status: 404 })
    }

    reportsStore.splice(index, 1)
    return new HttpResponse(null, { status: 204 })
  }),

  // Download report
  http.get(`${API_BASE}/reports/history/:reportId/download`, async ({ params }) => {
    await delay(500)
    const { reportId } = params
    const report = reportsStore.find((r) => r.id === reportId)

    if (!report) {
      return HttpResponse.json({ detail: 'Report not found' }, { status: 404 })
    }

    if (report.status !== 'completed') {
      return HttpResponse.json({ detail: `Report is not ready (status: ${report.status})` }, { status: 400 })
    }

    // Increment download count
    report.downloaded_count += 1
    report.last_downloaded_at = new Date().toISOString()

    // Return mock content based on format
    const contentMap: Record<string, { content: string; type: string }> = {
      html: { content: '<html><body><h1>Mock Report</h1></body></html>', type: 'text/html' },
      pdf: { content: 'Mock PDF content', type: 'application/pdf' },
      csv: { content: 'col1,col2,col3\nval1,val2,val3', type: 'text/csv' },
      json: { content: JSON.stringify({ report: 'Mock data' }), type: 'application/json' },
      markdown: { content: '# Mock Report\n\nContent here', type: 'text/markdown' },
      junit: { content: '<?xml version="1.0"?><testsuite></testsuite>', type: 'application/xml' },
      excel: { content: 'Mock Excel content', type: 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet' },
    }

    const { content, type } = contentMap[report.format] || contentMap.html

    return new HttpResponse(content, {
      headers: {
        'Content-Type': type,
        'Content-Disposition': `attachment; filename="${report.name}.${report.format}"`,
      },
    })
  }),

  // Generate report content
  http.post(`${API_BASE}/reports/history/:reportId/generate`, async ({ params }) => {
    await delay(1000) // Simulate generation time
    const { reportId } = params
    const index = reportsStore.findIndex((r) => r.id === reportId)

    if (index === -1) {
      return HttpResponse.json({ detail: 'Report not found' }, { status: 404 })
    }

    const report = reportsStore[index]

    if (report.status === 'completed') {
      return HttpResponse.json({ detail: 'Report already generated' }, { status: 400 })
    }

    // Update report to completed
    reportsStore[index] = {
      ...report,
      status: 'completed',
      file_path: `data/reports/${report.id}.${report.format}`,
      file_size: Math.floor(Math.random() * 1000000) + 1024,
      content_hash: Array.from({ length: 64 }, () => Math.floor(Math.random() * 16).toString(16)).join(''),
      generation_time_ms: Math.random() * 2000 + 100,
      download_url: `/api/v1/reports/history/${report.id}/download`,
      updated_at: new Date().toISOString(),
    }

    return HttpResponse.json(reportsStore[index])
  }),

  // Cleanup expired reports
  http.post(`${API_BASE}/reports/history/cleanup`, async () => {
    await delay(500)
    const before = reportsStore.length
    reportsStore = reportsStore.filter((r) => r.status !== 'expired')
    const deleted = before - reportsStore.length

    return HttpResponse.json({ deleted })
  }),

  // Bulk generate reports
  http.post(`${API_BASE}/reports/bulk`, async ({ request }) => {
    await delay(1500)
    const data = (await request.json()) as {
      validation_ids: string[]
      format?: string
      theme?: string
      locale?: string
      reporter_id?: string
      config?: Record<string, unknown>
      save_to_history?: boolean
      expires_in_days?: number
    }

    const reports: GeneratedReport[] = []
    const errors: Array<{ validation_id: string; error: string }> = []

    for (const validationId of data.validation_ids) {
      // Simulate occasional failures
      if (Math.random() < 0.1) {
        errors.push({
          validation_id: validationId,
          error: 'Validation not found or inaccessible',
        })
        continue
      }

      const report = createGeneratedReport({
        validation_id: validationId,
        format: (data.format as GeneratedReport['format']) || 'html',
        theme: data.theme || 'professional',
        locale: data.locale || 'en',
        reporter_id: data.reporter_id,
        config: data.config,
        status: 'completed',
        file_size: Math.floor(Math.random() * 1000000) + 1024,
        generation_time_ms: Math.random() * 2000 + 100,
      })

      if (data.save_to_history !== false) {
        reportsStore.unshift(report)
      }
      reports.push(report)
    }

    return HttpResponse.json({
      total: data.validation_ids.length,
      successful: reports.length,
      failed: errors.length,
      reports,
      errors,
    })
  }),
]

File: src/truthound_dashboard/core/report_history.py
"""Report History Service.

This module provides services for managing generated report history,
including CRUD operations, statistics, and cleanup.
"""

from __future__ import annotations

import hashlib
import os
from datetime import datetime, timedelta
from pathlib import Path
from typing import Any, Literal

from sqlalchemy import and_, desc, func, or_, select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from truthound_dashboard.db.models import (
    CustomReporter,
    GeneratedReport,
    ReportFormatType,
    ReportStatus,
    Source,
    Validation,
)


class ReportHistoryService:
    """Service for managing generated report history.

    Provides CRUD operations for report history records,
    statistics aggregation, and cleanup functionality.
    """

    def __init__(self, session: AsyncSession) -> None:
        """Initialize service with database session.

        Args:
            session: Async database session.
        """
        self.session = session
        self._reports_dir = Path("data/reports")

    async def list_reports(
        self,
        source_id: str | None = None,
        validation_id: str | None = None,
        reporter_id: str | None = None,
        format: str | None = None,
        status: str | None = None,
        include_expired: bool = False,
        search: str | None = None,
        sort_by: str = "created_at",
        sort_order: Literal["asc", "desc"] = "desc",
        page: int = 1,
        page_size: int = 20,
    ) -> tuple[list[GeneratedReport], int]:
        """List generated reports with filtering and pagination.

        Args:
            source_id: Filter by source ID.
            validation_id: Filter by validation ID.
            reporter_id: Filter by reporter ID.
            format: Filter by format.
            status: Filter by status.
            include_expired: Include expired reports.
            search: Search by name.
            sort_by: Field to sort by.
            sort_order: Sort direction.
            page: Page number (1-based).
            page_size: Items per page.

        Returns:
            Tuple of (reports list, total count).
        """
        # Build base query with relationships
        query = select(GeneratedReport).options(
            selectinload(GeneratedReport.source),
            selectinload(GeneratedReport.validation),
            selectinload(GeneratedReport.reporter),
        )

        # Apply filters
        conditions = []

        if source_id:
            conditions.append(GeneratedReport.source_id == source_id)

        if validation_id:
            conditions.append(GeneratedReport.validation_id == validation_id)

        if reporter_id:
            conditions.append(GeneratedReport.reporter_id == reporter_id)

        if format:
            try:
                format_enum = ReportFormatType(format.lower())
                conditions.append(GeneratedReport.format == format_enum)
            except ValueError:
                pass

        if status:
            try:
                status_enum = ReportStatus(status.lower())
                conditions.append(GeneratedReport.status == status_enum)
            except ValueError:
                pass

        if not include_expired:
            conditions.append(
                or_(
                    GeneratedReport.expires_at.is_(None),
                    GeneratedReport.expires_at > datetime.utcnow(),
                )
            )

        if search:
            conditions.append(GeneratedReport.name.ilike(f"%{search}%"))

        if conditions:
            query = query.where(and_(*conditions))

        # Get total count
        count_query = select(func.count()).select_from(
            query.subquery()
        )
        total = await self.session.scalar(count_query) or 0

        # Apply sorting
        sort_column = getattr(GeneratedReport, sort_by, GeneratedReport.created_at)
        if sort_order == "desc":
            query = query.order_by(desc(sort_column))
        else:
            query = query.order_by(sort_column)

        # Apply pagination
        offset = (page - 1) * page_size
        query = query.offset(offset).limit(page_size)

        result = await self.session.execute(query)
        reports = list(result.scalars().all())

        return reports, total

    async def get_report(self, report_id: str) -> GeneratedReport | None:
        """Get a single report by ID.

        Args:
            report_id: Report ID.

        Returns:
            GeneratedReport or None if not found.
        """
        query = (
            select(GeneratedReport)
            .options(
                selectinload(GeneratedReport.source),
                selectinload(GeneratedReport.validation),
                selectinload(GeneratedReport.reporter),
            )
            .where(GeneratedReport.id == report_id)
        )
        result = await self.session.execute(query)
        return result.scalar_one_or_none()

    async def create_report(
        self,
        name: str,
        format: str,
        validation_id: str | None = None,
        source_id: str | None = None,
        reporter_id: str | None = None,
        description: str | None = None,
        theme: str | None = None,
        locale: str = "en",
        config: dict[str, Any] | None = None,
        metadata: dict[str, Any] | None = None,
        expires_in_days: int | None = 30,
    ) -> GeneratedReport:
        """Create a new report record.

        Args:
            name: Report name.
            format: Report format.
            validation_id: Associated validation ID.
            source_id: Associated source ID.
            reporter_id: Custom reporter ID.
            description: Report description.
            theme: Visual theme.
            locale: Language locale.
            config: Generation config.
            metadata: Additional metadata.
            expires_in_days: Days until expiration.

        Returns:
            Created GeneratedReport.
        """
        # Calculate expiration
        expires_at = None
        if expires_in_days:
            expires_at = datetime.utcnow() + timedelta(days=expires_in_days)

        # Parse format
        try:
            format_enum = ReportFormatType(format.lower())
        except ValueError:
            format_enum = ReportFormatType.HTML

        report = GeneratedReport(
            name=name,
            description=description,
            format=format_enum,
            theme=theme,
            locale=locale,
            status=ReportStatus.PENDING,
            config=config,
            metadata=metadata,
            validation_id=validation_id,
            source_id=source_id,
            reporter_id=reporter_id,
            expires_at=expires_at,
        )

        self.session.add(report)
        await self.session.commit()
        await self.session.refresh(report)

        return report

    async def update_report(
        self,
        report_id: str,
        name: str | None = None,
        description: str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> GeneratedReport | None:
        """Update a report record.

        Args:
            report_id: Report ID.
            name: New name.
            description: New description.
            metadata: New metadata.

        Returns:
            Updated GeneratedReport or None if not found.
        """
        report = await self.get_report(report_id)
        if not report:
            return None

        if name is not None:
            report.name = name
        if description is not None:
            report.description = description
        if metadata is not None:
            report.metadata = metadata

        await self.session.commit()
        await self.session.refresh(report)
        return report

    async def delete_report(self, report_id: str) -> bool:
        """Delete a report record and its file.

        Args:
            report_id: Report ID.

        Returns:
            True if deleted, False if not found.
        """
        report = await self.get_report(report_id)
        if not report:
            return False

        # Delete file if exists
        if report.file_path and os.path.exists(report.file_path):
            try:
                os.remove(report.file_path)
            except OSError:
                pass

        await self.session.delete(report)
        await self.session.commit()
        return True

    async def mark_generating(self, report_id: str) -> GeneratedReport | None:
        """Mark report as generating.

        Args:
            report_id: Report ID.

        Returns:
            Updated report or None.
        """
        report = await self.get_report(report_id)
        if not report:
            return None

        report.status = ReportStatus.GENERATING
        await self.session.commit()
        await self.session.refresh(report)
        return report

    async def mark_completed(
        self,
        report_id: str,
        content: bytes | str,
        generation_time_ms: float,
    ) -> GeneratedReport | None:
        """Mark report as completed and store content.

        Args:
            report_id: Report ID.
            content: Report content.
            generation_time_ms: Generation time in milliseconds.

        Returns:
            Updated report or None.
        """
        report = await self.get_report(report_id)
        if not report:
            return None

        # Convert string to bytes if needed
        if isinstance(content, str):
            content = content.encode("utf-8")

        # Calculate hash
        content_hash = hashlib.sha256(content).hexdigest()

        # Store file
        self._reports_dir.mkdir(parents=True, exist_ok=True)

        # Determine extension
        ext_map = {
            ReportFormatType.HTML: ".html",
            ReportFormatType.CSV: ".csv",
            ReportFormatType.JSON: ".json",
            ReportFormatType.EXCEL: ".xlsx",
            ReportFormatType.CUSTOM: ".txt",
        }
        ext = ext_map.get(report.format, ".html")
        file_name = f"{report.id}{ext}"
        file_path = self._reports_dir / file_name

        with open(file_path, "wb") as f:
            f.write(content)

        # Update report
        report.mark_completed(
            file_path=str(file_path),
            file_size=len(content),
            generation_time_ms=generation_time_ms,
        )
        report.content_hash = content_hash

        await self.session.commit()
        await self.session.refresh(report)
        return report

    async def mark_failed(
        self,
        report_id: str,
        error_message: str,
    ) -> GeneratedReport | None:
        """Mark report as failed.

        Args:
            report_id: Report ID.
            error_message: Error message.

        Returns:
            Updated report or None.
        """
        report = await self.get_report(report_id)
        if not report:
            return None

        report.mark_failed(error_message)
        await self.session.commit()
        await self.session.refresh(report)
        return report

    async def record_download(self, report_id: str) -> GeneratedReport | None:
        """Record a download event.

        Args:
            report_id: Report ID.

        Returns:
            Updated report or None.
        """
        report = await self.get_report(report_id)
        if not report:
            return None

        report.increment_download()
        await self.session.commit()
        await self.session.refresh(report)
        return report

    async def get_report_content(self, report_id: str) -> tuple[bytes | None, str | None]:
        """Get report file content.

        Args:
            report_id: Report ID.

        Returns:
            Tuple of (content, content_type) or (None, None) if not found.
        """
        report = await self.get_report(report_id)
        if not report or not report.file_path:
            return None, None

        if not os.path.exists(report.file_path):
            return None, None

        content_type_map = {
            ReportFormatType.HTML: "text/html",
            ReportFormatType.CSV: "text/csv",
            ReportFormatType.JSON: "application/json",
            ReportFormatType.EXCEL: "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
            ReportFormatType.CUSTOM: "text/plain",
        }

        with open(report.file_path, "rb") as f:
            content = f.read()

        content_type = content_type_map.get(report.format, "application/octet-stream")
        return content, content_type

    async def get_statistics(self) -> dict[str, Any]:
        """Get report statistics.

        Returns:
            Statistics dictionary.
        """
        # Total reports
        total_query = select(func.count(GeneratedReport.id))
        total = await self.session.scalar(total_query) or 0

        # Total size
        size_query = select(func.coalesce(func.sum(GeneratedReport.file_size), 0))
        total_size = await self.session.scalar(size_query) or 0

        # Reports by format
        format_query = select(
            GeneratedReport.format,
            func.count(GeneratedReport.id),
        ).group_by(GeneratedReport.format)
        format_result = await self.session.execute(format_query)
        by_format = {
            row[0].value if hasattr(row[0], "value") else row[0]: row[1]
            for row in format_result
        }

        # Reports by status
        status_query = select(
            GeneratedReport.status,
            func.count(GeneratedReport.id),
        ).group_by(GeneratedReport.status)
        status_result = await self.session.execute(status_query)
        by_status = {
            row[0].value if hasattr(row[0], "value") else row[0]: row[1]
            for row in status_result
        }

        # Total downloads
        downloads_query = select(
            func.coalesce(func.sum(GeneratedReport.downloaded_count), 0)
        )
        total_downloads = await self.session.scalar(downloads_query) or 0

        # Average generation time
        avg_time_query = select(
            func.avg(GeneratedReport.generation_time_ms)
        ).where(GeneratedReport.generation_time_ms.isnot(None))
        avg_time = await self.session.scalar(avg_time_query)

        # Expired count
        expired_query = select(func.count(GeneratedReport.id)).where(
            and_(
                GeneratedReport.expires_at.isnot(None),
                GeneratedReport.expires_at < datetime.utcnow(),
            )
        )
        expired_count = await self.session.scalar(expired_query) or 0

        # Unique reporters used
        reporters_query = select(
            func.count(func.distinct(GeneratedReport.reporter_id))
        ).where(GeneratedReport.reporter_id.isnot(None))
        reporters_used = await self.session.scalar(reporters_query) or 0

        return {
            "total_reports": total,
            "total_size_bytes": int(total_size),
            "reports_by_format": by_format,
            "reports_by_status": by_status,
            "total_downloads": int(total_downloads),
            "avg_generation_time_ms": float(avg_time) if avg_time else None,
            "expired_count": expired_count,
            "reporters_used": reporters_used,
        }

    async def cleanup_expired(self) -> int:
        """Delete expired reports.

        Returns:
            Number of reports deleted.
        """
        # Find expired reports
        query = select(GeneratedReport).where(
            and_(
                GeneratedReport.expires_at.isnot(None),
                GeneratedReport.expires_at < datetime.utcnow(),
            )
        )
        result = await self.session.execute(query)
        expired = list(result.scalars().all())

        count = 0
        for report in expired:
            # Delete file
            if report.file_path and os.path.exists(report.file_path):
                try:
                    os.remove(report.file_path)
                except OSError:
                    pass

            await self.session.delete(report)
            count += 1

        await self.session.commit()
        return count

    async def find_by_hash(self, content_hash: str) -> GeneratedReport | None:
        """Find report by content hash for deduplication.

        Args:
            content_hash: Content hash.

        Returns:
            Existing report with same hash or None.
        """
        query = (
            select(GeneratedReport)
            .where(
                and_(
                    GeneratedReport.content_hash == content_hash,
                    GeneratedReport.status == ReportStatus.COMPLETED,
                    or_(
                        GeneratedReport.expires_at.is_(None),
                        GeneratedReport.expires_at > datetime.utcnow(),
                    ),
                )
            )
            .order_by(desc(GeneratedReport.created_at))
            .limit(1)
        )
        result = await self.session.execute(query)
        return result.scalar_one_or_none()
